rcore-fs-sefs/src/lib.rs

```diff
@@ -207,6 +210,131 @@ impl INodeImpl {
+            .checked_add(insert_range_len)
+            .ok_or(FsError::FileTooBig)?;
+        // Shift upward the contents of file starting at start_offset
+        let mut buf = unsafe { Box::<[u8; BLKSIZE]>::new_uninit().assume_init() };
+        let mut len = file_size - start_offset;
+        let mut offset = file_size;
+        while len != 0 {
+            let len_per_loop = len.min(BLKSIZE);
+            offset -= len_per_loop;
+            let len_read = self
+                .file
+                .read_at(&mut buf[..len_per_loop], offset)
+                .expect("failed to read");
+            assert!(len_read == len_per_loop);
+            let len_written = self
+                .file
+                .write_at(&buf[..len_per_loop], offset + insert_range_len)
+                .expect("failed to write");
+            assert!(len_written == len_per_loop);
+            len -= len_per_loop;
+        }
+        // Insert zeros
+        len = insert_range_len;
+        offset = start_offset;
+        static ZEROS: [u8; BLKSIZE] = [0; BLKSIZE];
+        while len != 0 {
+            let len_per_loop = len.min(BLKSIZE);
+            let len_written = self
+                .file
+                .write_at(&ZEROS[..len_per_loop], offset)
+                .expect("failed to write zeros");
+            len -= len_written;
+            offset += len_written;
+        }
+        // Update file size
+        self.file.set_len(new_file_size)?;
+        inode.size = new_file_size as u64;
+        Ok(())
+    }
+
+    fn collapse_range(&self, start_offset: usize, end_offset: usize) -> vfs::Result<()> {
+        assert!(start_offset < end_offset);
+        let mut inode = self.disk_inode.write();
+        let file_size = inode.size as usize;
+        if end_offset >= file_size {
+            return Err(FsError::InvalidParam);
+        }
+        let collapse_range_len = end_offset - start_offset;
+        if start_offset % BLKSIZE != 0 || collapse_range_len % BLKSIZE != 0 {
+            return Err(FsError::InvalidParam);
+        }
+        // Move the contents of file starting at end_offset to start_offset
+        let mut buf = unsafe { Box::<[u8; BLKSIZE]>::new_uninit().assume_init() };
+        let mut len = file_size - end_offset;
+        let mut read_offset = end_offset;
+        let mut write_offset = start_offset;
+        while len != 0 {
+            let len_per_loop = len.min(BLKSIZE);
+            let len_read = self
+                .file
+                .read_at(&mut buf[..len_per_loop], read_offset)
+                .expect("failed to read");
+            assert!(len_read == len_per_loop);
+            let len_written = self
+                .file
+                .write_at(&buf[..len_per_loop], write_offset)
+                .expect("failed to write");
+            assert!(len_written == len_per_loop);
+            len -= len_per_loop;
+            read_offset += len_per_loop;
+            write_offset += len_per_loop;
+        }
```

Comment:
Could you try to unify the two pieces of copying logic in `insert_range` and `collapse_range` by adding something like a `copy_range`? There is a bit of tricky in copying (you don't want overwrite the data that you haven't copied yet). By writing a separate function for this purpose, we can reduce code redundancy, make both `insert_range` and `collapse_range` methods simpler, and emphasize the tricky part of copying in the right place (the `copy_range` method).

Reply:
Done